pkg/resource/pipe/hooks.go

```diff
@@ -0,0 +1,990 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package pipe
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	svcsdk "github.com/aws/aws-sdk-go/service/pipes"
+
+	svcapitypes "github.com/aws-controllers-k8s/pipes-controller/apis/v1alpha1"
+)
+
+const (
+	defaultRequeueDelay = time.Second * 5
+)
+
+var (
+	requeueWaitWhileCreating = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, requeueing", svcsdk.PipeStateCreating),
+		defaultRequeueDelay,
+	)
+
+	requeueWaitWhileUpdating = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, cannot be modified or deleted", svcsdk.PipeStateUpdating),
+		defaultRequeueDelay,
+	)
+
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, cannot be modified or deleted", svcsdk.PipeStateDeleting),
+		defaultRequeueDelay,
+	)
+)
+
+// hasNilDifference returns true if the supplied subjects' nilness is
+// different
+func hasNilDifference(a, b interface{}) bool {
+	if isEmpty(a) || isEmpty(b) {
+		if (isEmpty(a) && isNotEmpty(b)) || (isEmpty(b) && isNotEmpty(a)) {
+			return true
+		}
+	}
+	return false
+}
+
+// isEmpty checks the passed interface argument for Nil or empty struct value
+// (with zero values). For interfaces, only 'i==nil' check is not sufficient.
+// https://tour.golang.org/methods/12 More details:
+// https://mangatmodi.medium.com/go-check-nil-interface-the-right-way-d142776edef1
+func isEmpty(i interface{}) bool {
+	if i == nil {
+		return true
+	}
+
+	switch reflect.TypeOf(i).Kind() {
+	case reflect.Ptr, reflect.Map, reflect.Array, reflect.Chan, reflect.Slice:
+		if reflect.ValueOf(i).IsNil() {
+			return true
+		}
+
+		if reflect.ValueOf(i).Elem().Kind() == reflect.Struct {
+			return reflect.ValueOf(i).Elem().IsZero()
+		}
+	}
+	return false
+}
+
+func isNotEmpty(i interface{}) bool {
+	return !isEmpty(i)
+}
+
+func customPreCompare(
+	delta *ackcompare.Delta,
+	a *resource,
+	b *resource,
+) {
+	aDesiredState := a.ko.Spec.DesiredState
+	bDesiredState := b.ko.Spec.DesiredState
+
+	// assumes API always returns desiredState
+	running := aDesiredState == nil || *aDesiredState == "" || *aDesiredState == svcsdk.PipeStateRunning
+	if !(running && *bDesiredState == svcsdk.PipeStateRunning) {
+		if *aDesiredState != *bDesiredState {
+			delta.Add("Spec.DesiredState", aDesiredState, bDesiredState)
+		}
+	}
+
+	// hack (by the great @a-hilaly): forces a requeue in update if currentState != desiredState to reconcile
+	// and update status fields in Kubernetes resource e.g., in case of UPDATE_FAILED
+	if *bDesiredState != *b.ko.Status.CurrentState {
+		// setting Spec. because Status. is not considered in delta logic
+		delta.Add("Spec.CurrentState", *bDesiredState, *b.ko.Status.CurrentState)
+	}
```

Comment:
Instead of doing this, you could rename the `DesiredState` field to just `State`, then ignore the `CreatePipeOutput.CurrentState` field path and in the `sdk_read_one_post_set_output` code hook, simply do:

```go
if resp.CurrentState != nil {
    ko.Spec.State = resp.CurrentState
}
```

then the standard comparison logic would be just fine.

Reply:
A better hack by the great James "Jay" Pipes

Reply:
The only concern I have here is from a debugging pov since an operator could be confused a) by the field name and b) why this field name has a value which is not allowed as per API

[DesiredState](https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_CreatePipe.html#API_CreatePipe_RequestSyntax)
The state the pipe should be in.

Type: String

Valid Values: RUNNING | STOPPED

Required: No

Reply:
@embano1 @A-Hilaly OK, I didn't realize the value set of these fields was different. I see now that this is more of a way for EventBridge to avoid having a `StopPipe` API call that only changes the desired state of the pipe. Alright, then, good to go with this :)